# PDAM TEST VER: 01

# 18/09/2023

- model_pdam v1.0.1 -> 1.0.2

Outline:

- add key tag bln ini to model

# 14/9/2023

- model_pdam v1.0.0 -> v1.0.1

Outline:

- change total tagihan to bill before tax, etc [Fixed]

# 13/9/2023

- NER V1.0.0 -> V1.0.1
- OCR V1.0.0 -> V1.0.0

Outline:

- fill pemakaian 2 hasilnya salah padahal awal dan akhir sudah benar, idx: 33 [Status: Fixed]
- bill harus dihilangkan angka yg desimalnya, contoh IDR 691,440.00' jadi 691440, idx: 8, 9 [Status: Fixed]
- improved levenshtein untuk key yang memiliki value number, (r"F:\Project_Base\oneapps\Backend\Vision\test_images\pdam\13-9-2023\*") idx: 2 [Status: Fixed]

# IMPORT

In [7]:
import os
os.chdir(r'F:\Project_Base\oneapps\Backend\Vision')

import model_predict
import spacy
import time
import requests
import json
import glob
import pandas as pd
from tqdm.notebook import tqdm

def load_model():
    start_time = time.time()
    model_folder = r'F:\Project_Base\oneapps\Backend\Vision\data\model_ner\model_pdam_1p0p2'
    nlp = spacy.load(model_folder)
    finish_time_load_model = (time.time() - start_time)
    print(f"time load model: {finish_time_load_model}")                                                                          
    return nlp

def predict(img_path, nlp):
    result = model_predict.ner_pdam(nlp, img_file=img_path, debug=False, line_sensitivity=0.2)
    return result

def predict_api(img_path):
    url = "http://127.0.0.1/predict_pdam"
    with open(img_path, 'rb') as img_file:
        files = {"image" : img_file}
        response = requests.post(url, files=files)
    return json.loads(response.text)

# LOAD MODEL

In [8]:
# nlp = load_model()

# INPUT LOCATION

In [9]:
# for test entire image inside folder
# img_path = glob.glob(r"F:\NawaAI\Notebooks\DEV_NER\PDAM\data\struk_pdam_used\*\*") # masukan folder di sini
img_path = glob.glob(r"F:\Project_Base\oneapps\Backend\Vision\test_script\test_images\pdam\21-9-2023\*")

# for test selected image
# img_path = [r'F:\Project_Base\oneapps\Backend\Vision\test_images\pdam\PUL.JPG', r'F:\NawaAI\Notebooks\DEV_NER\PDAM\data\struk_pdam_used\pos\KBM.JPG'] # masukan individual file di sini

# pilih tipe testnya, langsung nembak ke modulnya, atau ke api nya
# test = 'modul'
test = 'api'

In [10]:
# STOPP

# MODUL & API TEST

In [11]:
# untuk menyimpan semua hasil
result_all = []
# img_path = [img_path[18]]
for i, img in enumerate(tqdm(img_path)):
    print(i, img)

    if test == 'api':
        result = predict_api(img)
    elif test == 'modul':
        result = predict(img, nlp)
        
    display(result)
    result_all.append(result)

  0%|          | 0/3 [00:00<?, ?it/s]

0 F:\Project_Base\oneapps\Backend\Vision\test_script\test_images\pdam\21-9-2023\0.png


{'bill': '200000',
 'billingtype': 'PDAM_Postpaid',
 'custaddress': 'JL . PEMLIDA SELATAN NO.34',
 'custname': 'LUKITO SINDORO',
 'customerId': '07-16-27007',
 'listTransaction': [],
 'merchant': 'POS INDONESIA',
 'meter_akhir': '4784',
 'meter_awal': '4784',
 'meter_awal_akhir': [],
 'pdamname': 'PDAM KAB . KLATEN',
 'periode': '01-09-2022 00:00:00',
 'product': 'Water',
 'quantity': '0',
 'rate': '',
 'transactiondate': '05-10-2022 09:38:30',
 'transactiontype': 'WATER',
 'type': '209.000'}

1 F:\Project_Base\oneapps\Backend\Vision\test_script\test_images\pdam\21-9-2023\1.png


{'bill': '200000',
 'billingtype': 'PDAM_Postpaid',
 'custaddress': 'JL . PEMLIDA SELATAN NO.34',
 'custname': 'LUKITO SINDORO',
 'customerId': '07-16-27007',
 'listTransaction': [],
 'merchant': 'POS INDONESIA',
 'meter_akhir': '4784',
 'meter_awal': '4784',
 'meter_awal_akhir': [],
 'pdamname': 'PDAM KAB . KLATEN',
 'periode': '01-09-2022 00:00:00',
 'product': 'Water',
 'quantity': '0',
 'rate': '',
 'transactiondate': '05-10-2022 09:38:30',
 'transactiontype': 'WATER',
 'type': ''}

2 F:\Project_Base\oneapps\Backend\Vision\test_script\test_images\pdam\21-9-2023\2.png


{'bill': '200000',
 'billingtype': 'PDAM_Postpaid',
 'custaddress': 'JL . PEMLIDA SELATAN NO.34',
 'custname': 'LUKITO SINDORO',
 'customerId': '07-16-27007',
 'listTransaction': [],
 'merchant': 'POS INDONESIA',
 'meter_akhir': '4784',
 'meter_awal': '4784',
 'meter_awal_akhir': [],
 'pdamname': 'PDAM KAB . KLATEN',
 'periode': '01-09-2022 00:00:00',
 'product': 'Water',
 'quantity': '0',
 'rate': '',
 'transactiondate': '05-10-2022 09:38:30',
 'transactiontype': 'WATER',
 'type': ''}

# STOPPER

In [12]:
STOPPP

NameError: name 'STOPPP' is not defined

# SHOW NULL SUMMARY

In [ ]:
import sys
sys.getsizeof("")

51

In [ ]:
null_value = 0
all_data = 0
null_detail = {'transactiondate': 0,
 'custname': 0,
 'custaddress': 0,
 'type': 0,
 'pdamname': 0,
 'periode': 0,
 'quantity': 0,
 'bill': 0,
 'customerId': 0,
 'meter_awal': 0,
 'meter_akhir': 0,
 'meter_awal_akhir': 0,
 'merchant': 0,
 'billingtype': 0,
 'listTransaction': 0,
 'transactiontype': 0,
 'product': 0,
 'rate': 0}

for result in result_all:
    for ent in result:
        if result[ent] == '':
            null_value += 1
        if not result[ent]:
            null_detail[ent] += 1
        all_data += 1

print(f"jumlah all data: {all_data}")
print(f"jumlah null: {null_value}")
display("null detail :", null_detail)

df = pd.DataFrame(null_detail, index=['NER'])
display(df)

jumlah all data: 54
jumlah null: 5


'null detail :'

{'transactiondate': 0,
 'custname': 0,
 'custaddress': 0,
 'type': 2,
 'pdamname': 0,
 'periode': 0,
 'quantity': 0,
 'bill': 0,
 'customerId': 0,
 'meter_awal': 0,
 'meter_akhir': 0,
 'meter_awal_akhir': 3,
 'merchant': 0,
 'billingtype': 0,
 'listTransaction': 3,
 'transactiontype': 0,
 'product': 0,
 'rate': 3}

transactiondate  custname  custaddress  type  pdamname  periode  \
NER                0         0            0     2         0        0   

     quantity  bill  customerId  meter_awal  meter_akhir  meter_awal_akhir  \
NER         0     0           0           0            0                 3   

     merchant  billingtype  listTransaction  transactiontype  product  rate  
NER         0            0                3                0        0     3

# SHOW INDIVIDUAL VALUE

In [ ]:
values = {'transactiondate': [],
 'custname': [],
 'custaddress': [],
 'type': [],
 'pdamname': [],
 'periode': [],
 'quantity': [],
 'bill': [],
 'customerId': [],
 'meter_awal': [],
 'meter_akhir': [],
 'meter_awal_akhir': [],
 'merchant': [],
 'billingtype': [],
 'listTransaction': [],
 'transactiontype': [],
 'product': [],
 'rate': []}

for res in result_all:
    try:
        for label in res:
            values[label].append(res[label])
    except:
        pass

for label in values:
    print(f'{label.upper()}:')
    for i, val in enumerate(values[label]):
        print(i, val, img_path[i])
    print()

TRANSACTIONDATE:
0 05-10-2022 09:38:30 F:\Project_Base\oneapps\Backend\Vision\test_script\test_images\pdam\21-9-2023\0.png
1 05-10-2022 09:38:30 F:\Project_Base\oneapps\Backend\Vision\test_script\test_images\pdam\21-9-2023\1.png
2 05-10-2022 09:38:30 F:\Project_Base\oneapps\Backend\Vision\test_script\test_images\pdam\21-9-2023\2.png

CUSTNAME:
0 LUKITO SINDORO F:\Project_Base\oneapps\Backend\Vision\test_script\test_images\pdam\21-9-2023\0.png
1 LUKITO SINDORO F:\Project_Base\oneapps\Backend\Vision\test_script\test_images\pdam\21-9-2023\1.png
2 LUKITO SINDORO F:\Project_Base\oneapps\Backend\Vision\test_script\test_images\pdam\21-9-2023\2.png

CUSTADDRESS:
0 JL . PEMLIDA SELATAN NO.34 F:\Project_Base\oneapps\Backend\Vision\test_script\test_images\pdam\21-9-2023\0.png
1 JL . PEMLIDA SELATAN NO.34 F:\Project_Base\oneapps\Backend\Vision\test_script\test_images\pdam\21-9-2023\1.png
2 JL . PEMLIDA SELATAN NO.34 F:\Project_Base\oneapps\Backend\Vision\test_script\test_images\pdam\21-9-2023\2.pn

# NULL VALUE LIST

## TIDAK ADA PERIODE

- 31: krn 'SEPTEMBER 20122'

In [ ]:
idx = [31]
for i in idx:
    print(img_path[i])
    if test == 'api':
        result = predict_api(img_path[i])
    elif test == 'modul':
        result = predict(img_path[i], nlp)
    display(result)
    print("====="*10)
    print()

IndexError: list index out of range

## TIDAK ADA QUANTITY

- 20: krn tidak ada meter akhir
- 22: krn tidak ada meter awal dan meter akhir
- 26: krn tidak ada meter awal dan meter akhir
- 36: krn tidak ada meter awal dan meter akhir

In [ ]:
idx = [20, 22, 26, 36]
for i in idx:
    print(i, img_path[i])
    if test == 'api':
        result = predict_api(img_path[i])
    elif test == 'modul':
        result = predict(img_path[i], nlp)
    display(result)
    print("====="*10)
    print()

20 F:\NawaAI\Notebooks\DEV_NER\PDAM\data\struk_pdam_used\pos\bgr.JPG


{'transactiondate': '08-08-2022 13:49:55',
 'custname': 'HAMA PT RAYA PERSADA INDAH O. WIDODO . DANIEL',
 'custaddress': 'JL IR H JUANDA 12',
 'type': '',
 'pdamname': '',
 'periode': '01-07-2022 00:00:00',
 'quantity': '',
 'bill': '655300',
 'customerId': '11501126',
 'meter_awal': '531',
 'meter_akhir': '',
 'meter_awal_akhir': [],
 'merchant': 'POS INDONESIA',
 'billingtype': 'PDAM_Postpaid',
 'listTransaction': [],
 'transactiontype': 'WATER',
 'product': 'Water',
 'rate': ''}


22 F:\NawaAI\Notebooks\DEV_NER\PDAM\data\struk_pdam_used\pos\BKT.JPG


{'transactiondate': '09-08-2022 00:00:00',
 'custname': 'PT.BANK NIAGA',
 'custaddress': '',
 'type': 'WK',
 'pdamname': '',
 'periode': '01-08-2022 00:00:00',
 'quantity': '',
 'bill': '6225431',
 'customerId': '0132170015',
 'meter_awal': '',
 'meter_akhir': '',
 'meter_awal_akhir': [],
 'merchant': 'POS INDONESIA',
 'billingtype': 'PDAM_Postpaid',
 'listTransaction': [],
 'transactiontype': 'WATER',
 'product': 'Water',
 'rate': ''}


26 F:\NawaAI\Notebooks\DEV_NER\PDAM\data\struk_pdam_used\pos\HGG.JPG


{'transactiondate': '07-10-2022 10:45:34',
 'custname': 'D KURNIAWAN',
 'custaddress': '',
 'type': '',
 'pdamname': 'PDAM DKI JAKARTA',
 'periode': '01-09-2022 00:00:00',
 'quantity': '',
 'bill': '120362',
 'customerId': '000277969',
 'meter_awal': '',
 'meter_akhir': '',
 'meter_awal_akhir': [],
 'merchant': 'POS INDONESIA',
 'billingtype': 'PDAM_Postpaid',
 'listTransaction': [],
 'transactiontype': 'WATER',
 'product': 'Water',
 'rate': ''}


36 F:\NawaAI\Notebooks\DEV_NER\PDAM\data\struk_pdam_used\pos\wbc.JPG


{'transactiondate': '09-08-2022 09:41',
 'custname': 'PT LIPPO BANK',
 'custaddress': '',
 'type': '',
 'pdamname': 'PDAM PALYJA JAKARTA',
 'periode': '01-07-2023 00:00:00',
 'quantity': '',
 'bill': '4982275',
 'customerId': '000314696',
 'meter_awal': '',
 'meter_akhir': '',
 'meter_awal_akhir': [],
 'merchant': 'POS INDONESIA',
 'billingtype': 'PDAM_Postpaid',
 'listTransaction': [],
 'transactiontype': 'WATER',
 'product': 'Water',
 'rate': ''}

## TIDAK ADA CUSTOMERID

- 8: krn tdk ada customerid
- 9: krn tdk ada customerid

In [ ]:
idx = [8, 9]
for i in idx:
    print(i, img_path[i])
    if test == 'api':
        result = predict_api(img_path[i])
    elif test == 'modul':
        result = predict(img_path[i], nlp)
    display(result)
    print("====="*10)
    print()

8 F:\NawaAI\Notebooks\DEV_NER\PDAM\data\struk_pdam_used\new_gomobile\GOMobile_2_page-0001.jpg


{'transactiondate': '15-02-2023 10:20',
 'custname': 'PT.SURABAYA LAND',
 'custaddress': '',
 'type': '',
 'pdamname': 'PDAM KOTA SURABAYA',
 'periode': '01-02-2023 00:00:00',
 'quantity': '59',
 'bill': '691440',
 'customerId': '',
 'meter_awal': '103',
 'meter_akhir': '162',
 'meter_awal_akhir': [103, 162],
 'merchant': 'OCTO',
 'billingtype': 'PDAM_Postpaid',
 'listTransaction': [],
 'transactiontype': 'WATER',
 'product': 'Water',
 'rate': ''}


9 F:\NawaAI\Notebooks\DEV_NER\PDAM\data\struk_pdam_used\new_gomobile\go_mobile_1.jpg


{'transactiondate': '31-05-2023 17:22',
 'custname': 'NAME JOHANES',
 'custaddress': '',
 'type': '',
 'pdamname': '',
 'periode': '01-05-2023 00:00:00',
 'quantity': '5',
 'bill': '115940',
 'customerId': '',
 'meter_awal': '7841',
 'meter_akhir': '7846',
 'meter_awal_akhir': [7841, 7846],
 'merchant': 'OCTO',
 'billingtype': 'PDAM_Postpaid',
 'listTransaction': [],
 'transactiontype': 'WATER',
 'product': 'Water',
 'rate': ''}

## TIDAK ADA MERCHANT

- 6: krn tdk ada merchant
- 17: krn tdk ada merchant
- 18: krn tdk ada merchant

In [ ]:
270 == 270.0

True

In [ ]:
idx = [6, 17, 18]
for i in idx:
    print(i, img_path[i])
    if test == 'api':
        result = predict_api(img_path[i])
    elif test == 'modul':
        result = predict(img_path[i], nlp)
    display(result)
    print("====="*10)
    print()

6 F:\NawaAI\Notebooks\DEV_NER\PDAM\data\struk_pdam_used\indomaret\TKS.JPG


{'transactiondate': '10-08-2022 08:48:01',
 'custname': 'LIM TONG CIL',
 'custaddress': 'ALARAT KISANUN 120',
 'type': '',
 'pdamname': 'PDAM KOTA GR',
 'periode': '01-07-2022 00:00:00',
 'quantity': '46',
 'bill': '437500',
 'customerId': '0952147967',
 'meter_awal': '3025',
 'meter_akhir': '3071',
 'meter_awal_akhir': [3025, 3071],
 'merchant': '',
 'billingtype': 'PDAM_Postpaid',
 'listTransaction': [],
 'transactiontype': 'WATER',
 'product': 'Water',
 'rate': ''}


17 F:\NawaAI\Notebooks\DEV_NER\PDAM\data\struk_pdam_used\new_tokopedia\TokPed_2_page-0001.jpg


{'transactiondate': '06-03-2023 11:04',
 'custname': 'CIMB NIAGA',
 'custaddress': 'CENTRAL EROPA BLOK G / 2',
 'type': 'N1',
 'pdamname': '',
 'periode': '01-02-2023 00:00:00',
 'quantity': '11',
 'bill': '127200',
 'customerId': '16029215',
 'meter_awal': '2159',
 'meter_akhir': '2170',
 'meter_awal_akhir': [],
 'merchant': '',
 'billingtype': 'PDAM_Postpaid',
 'listTransaction': [],
 'transactiontype': 'WATER',
 'product': 'Water',
 'rate': ''}


18 F:\NawaAI\Notebooks\DEV_NER\PDAM\data\struk_pdam_used\new_tokopedia\TokPed_2_page-0002.jpg


{'transactiondate': '06-03-2023 11:02',
 'custname': 'CIMB NIAGA',
 'custaddress': 'CENTRAL EROPA BLOK G / 1',
 'type': 'N1',
 'pdamname': '',
 'periode': '01-02-2023 00:00:00',
 'quantity': '5',
 'bill': '77700',
 'customerId': '16029115',
 'meter_awal': '950',
 'meter_akhir': '955',
 'meter_awal_akhir': [],
 'merchant': '',
 'billingtype': 'PDAM_Postpaid',
 'listTransaction': [],
 'transactiontype': 'WATER',
 'product': 'Water',
 'rate': ''}

# WRONG VALUE LIST

## SALAH QUANTITY

- 31: harusnya 277.440 krn key tag bln ini tdk ada textnya, maka akan mengambil ke key yang lain[Status: Unfixed krn yg salah dari struknya]
- 34: 839, meter akhir sudah bener 847, meter awal terbaca 8:32, ketika di clean jadi kebaca 8 aja, jadi hasilnya salah [Status: Unfixed krn yg salah dari struknya]

- 28: 536.800, krn pemakaian air tdk ada text nya [Status: Fixed, hasil jadi 113]
- 33: 1210, sudah bener awal dan akhir tapi pas fill pemakaian 2 hasilnya salah [Status: Fixed, hasil jadi 2]

In [ ]:
idx = [31, 34]
for i in idx:
    print(i, img_path[i])
    if test == 'api':
        result = predict_api(img_path[i])
    elif test == 'modul':
        result = predict(img_path[i], nlp)
    display(result)
    print("====="*10)
    print()

31 F:\NawaAI\Notebooks\DEV_NER\PDAM\data\struk_pdam_used\pos\RYD.JPG


{'transactiondate': '05-10-2022 09:38:26',
 'custname': 'PT BANK NIAGA',
 'custaddress': 'SLAMET RIYADI 8',
 'type': '0FC7720CC0FACBF6C34A8D767ECEEC6D PB',
 'pdamname': 'PERUNDA AIR MINUM KOTA SURAKARTA',
 'periode': '',
 'quantity': '0',
 'bill': '67500',
 'customerId': '00003263',
 'meter_awal': '385',
 'meter_akhir': '385',
 'meter_awal_akhir': [],
 'merchant': 'POS INDONESIA',
 'billingtype': 'PDAM_Postpaid',
 'listTransaction': [],
 'transactiontype': 'WATER',
 'product': 'Water',
 'rate': ''}


34 F:\NawaAI\Notebooks\DEV_NER\PDAM\data\struk_pdam_used\pos\STL.JPG


{'transactiondate': '08-08-2022 10:14:41',
 'custname': 'PT.BANK CIMB NIAGA TBK',
 'custaddress': 'XPNA -0001ELOK A NO.0001',
 'type': '',
 'pdamname': 'FDAM TIRTA KAHURIPAN',
 'periode': '01-07-2022 00:00:00',
 'quantity': '15',
 'bill': '201500',
 'customerId': '23100116',
 'meter_awal': '8',
 'meter_akhir': '847',
 'meter_awal_akhir': [],
 'merchant': 'POS INDONESIA',
 'billingtype': 'PDAM_Postpaid',
 'listTransaction': [],
 'transactiontype': 'WATER',
 'product': 'Water',
 'rate': ''}

## SALAH BILL

---
- 14: 2022, Pemakaian Air ( 8 ) : Rp 100.40 kena validasi krn angka salah [Unfixed, krn salah di struknya]
- 19: harusnya tag bln ini, krn Tag Bln Ini tdk ada nilainya jadi ekstraksi dari yg lain [Unfixed, krn salah di struknya]
---
- 31: 154000, jadi 67.500 [Fixed]
- 8: 69144000, krn bill diambil semuanya,  terdapat tanda (, dan .) yang dihapus, harusnya dari angka awal sampai titik saja ngambilnya [Fixed, jadi 691440]
- 9: 11594000, sama seperti diatas [Fixed, jadi 115940]


In [ ]:
idx = [14, 19]
for i in idx:
    print(i, img_path[i])
    if test == 'api':
        result = predict_api(img_path[i])
    elif test == 'modul':
        result = predict(img_path[i], nlp)
    display(result)
    print("====="*10)
    print()

14 F:\NawaAI\Notebooks\DEV_NER\PDAM\data\struk_pdam_used\new_lyonnaise\SPO.JPG


{'transactiondate': '24-08-2022 00:00:00',
 'custname': 'R DELTA IND LTD',
 'custaddress': 'TANGGAL HULAI BAYAR 27-08-2022',
 'type': '.  / KELOMP 3Q / K4B',
 'pdamname': 'PAM LYONNAISE JAYA',
 'periode': '01-08-2022 00:00:00',
 'quantity': '8',
 'bill': '2022',
 'customerId': '0209',
 'meter_awal': '8394',
 'meter_akhir': '8101',
 'meter_awal_akhir': [],
 'merchant': 'PALYJA',
 'billingtype': 'PDAM_Postpaid',
 'listTransaction': [],
 'transactiontype': 'WATER',
 'product': 'Water',
 'rate': ''}


19 F:\NawaAI\Notebooks\DEV_NER\PDAM\data\struk_pdam_used\pos\ADY.JPG


{'transactiondate': '05-09-2022 10:25:53',
 'custname': 'GATOT WIBISONO',
 'custaddress': 'JEMUR ANDAYANI 53 A',
 'type': '40',
 'pdamname': 'PDAM KOTA SURABAYA',
 'periode': '01-08-2022 00:00:00',
 'quantity': '34',
 'bill': '303440',
 'customerId': '1050029',
 'meter_awal': '171',
 'meter_akhir': '205',
 'meter_awal_akhir': [],
 'merchant': 'POS INDONESIA',
 'billingtype': 'PDAM_Postpaid',
 'listTransaction': [],
 'transactiontype': 'WATER',
 'product': 'Water',
 'rate': ''}

# KESIMPULAN YG HARUS DIPERBAIKI

# 13/9/2023

Fixable error:
- fill pemakaian 2 hasilnya salah padahal awal dan akhir sudah benar, idx: 33 [Status: Fixed]
- bill harus dihilangkan angka yg desimalnya, contoh IDR 691,440.00' jadi 691440, idx: 8, 9 [Status: Fixed]

# STOP

In [ ]:
# STOPPP

# FOR FIXING

## FIX FILL PEMAKAIAN 2 YG SALAH [FIXED]

In [ ]:
idx = [33]
for i in idx:
    print(i, img_path[i])
    if test == 'api':
        result = predict_api(img_path[i])
    elif test == 'modul':
        result = predict(img_path[i], nlp)
    display(result)
    print("====="*10)
    print()

33 F:\NawaAI\Notebooks\DEV_NER\PDAM\data\struk_pdam_used\pos\SKN.JPG


{'transactiondate': '08-08-2022 13:49:47',
 'custname': '12611363',
 'custaddress': 'JL SURYAKENCANA 174 A ** FERIODE 1.',
 'type': '',
 'pdamname': 'PDAM KOTA BOGOR',
 'periode': '01-07-2022 00:00:00',
 'quantity': '2',
 'bill': '81300',
 'customerId': '12611363',
 'meter_awal': '1210',
 'meter_akhir': '1212',
 'meter_awal_akhir': [],
 'merchant': 'POS INDONESIA',
 'billingtype': 'PDAM_Postpaid',
 'listTransaction': [],
 'transactiontype': 'WATER',
 'product': 'Water',
 'rate': ''}

## FIX ANGKA DESIMAL DI BILL [FIXED]

In [ ]:
text = "691,440.00"

def clean_currency_decimal(text):
    """
    cleaning currency
    example: IDR 691,440.00' -> 691,440
    """
    if "," in text and "." in text:
        comma_idx = text.find(",")
        period_idx = text.find(".")

        if period_idx > comma_idx:
            idx_decimal = period_idx
        else:
            idx_decimal = comma_idx
        
        if len(text[idx_decimal:]) == 3:
            text = text[:idx_decimal]

    return text

clean_currency_decimal(text)

'691,440'

In [ ]:
idx = [8, 9]
for i in idx:
    print(i, img_path[i])
    if test == 'api':
        result = predict_api(img_path[i])
    elif test == 'modul':
        result = predict(img_path[i], nlp)
    display(result)
    print("====="*10)
    print()

8 F:\NawaAI\Notebooks\DEV_NER\PDAM\data\struk_pdam_used\new_gomobile\GOMobile_2_page-0001.jpg


{'transactiondate': '15-02-2023 10:20',
 'custname': 'PT.SURABAYA LAND',
 'custaddress': '',
 'type': '',
 'pdamname': 'PDAM KOTA SURABAYA',
 'periode': '01-02-2023 00:00:00',
 'quantity': '59',
 'bill': '691440',
 'customerId': '',
 'meter_awal': '103',
 'meter_akhir': '162',
 'meter_awal_akhir': [103, 162],
 'merchant': 'OCTO',
 'billingtype': 'PDAM_Postpaid',
 'listTransaction': [],
 'transactiontype': 'WATER',
 'product': 'Water',
 'rate': ''}


9 F:\NawaAI\Notebooks\DEV_NER\PDAM\data\struk_pdam_used\new_gomobile\go_mobile_1.jpg


{'transactiondate': '31-05-2023 17:22',
 'custname': 'NAME JOHANES',
 'custaddress': '',
 'type': '',
 'pdamname': '',
 'periode': '01-05-2023 00:00:00',
 'quantity': '5',
 'bill': '115940',
 'customerId': '',
 'meter_awal': '7841',
 'meter_akhir': '7846',
 'meter_awal_akhir': [7841, 7846],
 'merchant': 'OCTO',
 'billingtype': 'PDAM_Postpaid',
 'listTransaction': [],
 'transactiontype': 'WATER',
 'product': 'Water',
 'rate': ''}

# TEXT LINE TEST

In [ ]:
# text_line = ['437980343', 'PT FOS INDONESIA ( PERSERO )', 'V', 'Kantor Kp . BOGOR 16000', 'TANDA TERIMA', 'ZAZ', 'PDAM KOTA BOGOR', 'Tanggal # 08-08-2022 13:49:55', 'No.Resi 16000-58 / 2022 / 815893 Petugas 550007815', 'NO PELANGGAN +11501126 Daniel', 'PT RAYA PERSADA INDAH DANTEL O. WIDODO .', 'ALAMAT JL IR H JUANDA 12 BOSM', 'PERIODE = JULI 2022 METERAI', 'METER AWAL 531 CICILAN AIR', '"', 'PETER AKHIR . 564 Rp .', 'TAGIHAN AIR Rp . 655.300 TOTAL TAGIHAN : Rp . 655.300', 'ADMIN FOS Rp . 2.500 %', 'TOTAL BAYAR : Rp . 657.800', 'MA', '( enam ratus lima puluh tujuh ribu delapan ratus Rupiah )', 'Bukti Pembayaran ini adalah SAH jika sudah dibubuhi CAP FUS', 'Bayarlah Tagihan PDAM Anda Sebelum TANTORIS', 'NESIA', 'NFUP 01.001.620.2-093.000 a.n PT Pos Indonesia', 'Syarat dan ketentuan berlaku Lačak status http://www.posindonesia.co.id', '1000 Al']
# text_line = ['PY COPY ( C437980343', 'PT FOS INDONESIA ( PERSERO )', 'V', 'Kantor Kp . BOGOR 16000', 'TANDA TERIMA', '402 .', 'PDAM KOTA BOGOR', 'Tanggal # 08-08-2022 13:49:55', 'No.Resi : 16000-58 / 2022 / 815893 Petugas 550007815', 'NO PELANGGAN +11501126 Daniel', 'PT RAYA PERSADA INDAH DANTEL O. WIDODO .', 'ALAMAT JL IR H JUANDA 12 BOSM', 'PERIODE ... = JULI 2022 METERAI .', 'METER AWAL 531 CICILAN AIR', '4', 'PETER AKHIR . 564 Rp .', 'TAGIHAN AIR Rp . 655.300 TOTAL TAGIHAN : Rp . 655.300', 'ADMIN FOS Rp . 2.500 %', 'TOTAL BAYAR : Rp . 657.800', '( enam ratus lima puluh tujuh ribu delapan ratus Rupiah )', 'Bukti Pembayaran ini adalah SAH jika sudah dibubuhi CAP FUS', 'A', 'Bayarlah Tagihan PDAM Anda Sebelum TANTORIS', 'INDONESIA', 'NFUP 01.001.620.2-093.000 a.n PT Pos Indonesia', 'Syarat dan ketentuan berlaku Lačak status http://www.posindonesia.co.id', '1000 Al']
text_line = ['PT POS INDONESIA ( PERSERO ) 9.017.579 295594543',
  'Kantor Kp . SURABAYA SELATAN 60300',
  'TANDA TERIMA',
  'PDAM KOTA SURABAYA',
  'Tanggal : 05-09-2022 10:25:53',
  'No.Resi : 60300-83 / 2022 / 815362 Petugas : 550007840',
  'No Pelanggan : 1050029',
  'Nama : GATOT WIBISONO',
  'Alamat : JEMUR ANDAYANI 53 A',
  'Gol Tarif : 40',
  'Hkm Reff : 20220905102439000000 63047464',
  'Bln / Thn : AGUSTUS 2022 Ret Kebersihan : Rp ..',
  'Feter Lalu : 171 Tunggakan Denda : Rp .',
  'Meter Kini : 205 Total Tagihan : Rp . 300.940',
  'Pemakaian : 34 Biaya Admin : Rp .',
  'Tag Bln Ini Rp . 505.000 Total Bayar : Rp . 303.440',
  '( tiga ratus tiga ribu empat ratus empat puluh Rupiah )',
  'Bukti pembayaran ini adalah sah jika sudah dibubuhi Cap Tanggal POS',
  'Terima kasih telah melakukan pemby tepat waktu',
  'KANTORPOS POS INDONESIA',
  'Syarat dan ketentuan berlaku Lacak status : http://www.posindonesia.co.id']
  
text = ' '.join(text_line)
doc = nlp(text.upper())
for ent in doc.ents:
    print(ent.label_, ":", ent.text)

MERCHANT : POS INDONESIA
NAMA_PDAM : PDAM KOTA SURABAYA
TANGGAL_PEMBAYARAN : 05-09-2022
NO_PELANGGAN : NO PELANGGAN : 1050029
NAMA_PELANGGAN : NAMA : GATOT WIBISONO
ALAMAT_PELANGGAN : ALAMAT : JEMUR ANDAYANI 53 A
GOLONGAN : GOL TARIF : 40
PERIODE : AGUSTUS 2022
METER_AWAL : FETER LALU : 171
METER_AKHIR : METER KINI : 205
PEMAKAIAN : PEMAKAIAN : 34
TOTAL_TAGIHAN : TAG BLN INI RP . 505.000
MERCHANT : POS INDONESIA


In [ ]:
doc = nlp('Gol . Tarif / Kelomp 3Q / K4B Pemakaian Air ( 8 ) : Rp 100.40')
for ent in doc.ents:
    print(f"{ent.label_}: {ent.text}")

GOLONGAN: Gol . Tarif / Kelomp 3Q / K4B
TOTAL_TAGIHAN: Pemakaian Air ( 8 ) : Rp 100.40


In [ ]:
from NawaAI.NER import postprocess
ents_dict = {}

text = "PEMAKAIAN AIR ( 8 ) : RP 100.400"
clean_total_tagihan = postprocess.get_number(text)
print(clean_total_tagihan)
if postprocess.currency_validator(clean_total_tagihan):
    clean_total_tagihan = postprocess.clean_currency_decimal(clean_total_tagihan)
    ents_dict["TOTAL_TAGIHAN"] = clean_total_tagihan.replace(",", "").replace(
        ".", ""
    )

print(ents_dict)

8
{}


In [ ]:
from NawaAI.NER import postprocess
text = "123-123"

postprocess.get_number(text)

'123-123'